gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,428 @@
+
+          DataNode currentNode = this.flowGraph.getNode(flowEdgeContext.getEdge().getEndPoints().get(1));
+          DatasetDescriptor currentOutputDatasetDescriptor = flowEdgeContext.getOutputDatasetDescriptor();
+
+          //Are we done?
+          if (isPathFound(currentNode, destNode, currentOutputDatasetDescriptor, destDatasetDescriptor)) {
+            return buildDag(flowEdgeContext);
+          }
+
+          //Expand the currentNode to its adjacent edges and add them to the queue.
+          List<FlowEdgeContext> nextEdges =
+              getNextEdges(currentNode, currentOutputDatasetDescriptor, destDatasetDescriptor, flowConfig);
+          for (FlowEdgeContext childFlowEdgeContext : nextEdges) {
+            //Add a pointer from the child edge to the parent edge, if the child edge is not already in the
+            // queue.
+            if (!this.pathMap.containsKey(childFlowEdgeContext)) {
+              edgeQueue.add(childFlowEdgeContext);
+              this.pathMap.put(childFlowEdgeContext, flowEdgeContext);
+            }
+          }
+        }
+      }
+      //No path found. Return an empty dag.
+      return new Dag<>(new ArrayList<>());
+    } catch (SpecNotFoundException | JobTemplate.TemplateException | IOException | URISyntaxException e) {
+      throw new PathFinderException(
+          "Exception encountered when computing path from src: " + srcNodeId + " to dest: " + destNodeId, e);
+    }
+  }
+
+  private boolean isPathFound(DataNode currentNode, DataNode destNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor) {
+    if ((currentNode.equals(destNode)) && (currentDatasetDescriptor.equals(destDatasetDescriptor))) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * A helper method that sorts the {@link FlowEdge}s incident on srcNode based on whether the FlowEdge has an
+   * output {@link DatasetDescriptor} that is compatible with the targetDatasetDescriptor.
+   * @param dataNode
+   * @param currentDatasetDescriptor Output {@link DatasetDescriptor} of the current edge.
+   * @param destDatasetDescriptor Target {@link DatasetDescriptor}.
+   * @return prioritized list of {@link FlowEdge}s to be added to the edge queue for expansion.
+   */
+  private List<FlowEdgeContext> getNextEdges(DataNode dataNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor, Config userConfig) {
+    List<FlowEdgeContext> prioritizedEdgeList = new LinkedList<>();
+    for (FlowEdge flowEdge : this.flowGraph.getEdges(dataNode)) {
+      try {
+        DataNode edgeDestination = this.flowGraph.getNode(flowEdge.getEndPoints().get(1));
+        //Should we skip this edge from any further consideration?
+        if (!edgeDestination.isActive() || !flowEdge.isActive()) {
+          continue;
+        }
+        for (Pair<DatasetDescriptor, DatasetDescriptor> datasetDescriptorPair : flowEdge.getFlowTemplate()
```

Comment:
If we moved this logic to `FlowTemplate` and expose it as a function, we could simplify `FlowGraphPathFinder` to some degree.

Reply:
I think it would be a bit too cumbersome to move this logic to FlowTemplate. Further, it would need the FlowTemplate to become aware of graph constructs such as FlowEdge, which does not seem ideal either.

Reply:
FlowEdge does not need to, but its merged config, be exposed to FlowTemplate